feat(device-service): CAN device service

vijay-hcl-epl opened the pull request:
Adding initial version of CAN device service.

lenny-intel commented on Makefile:
```diff
@@ -0,0 +1,39 @@
+
+build: ./VERSION ${MICROSERVICES}
+
+build/release/device-can-c/device-can-c:
```
```suggestion
build/release/device-can/device-can:
```

vijay-hcl-epl replied:
Corrected

lenny-intel commented on Makefile:
```diff
@@ -0,0 +1,39 @@
+MICROSERVICES=build/release/device-can-c/device-can-c
+.PHONY: $(MICROSERVICES)
+
+DOCKERS=docker_device_can_c
```
```suggestion
DOCKERS=docker_device_can
```

vijay-hcl-epl replied:
Corrected

lenny-intel commented on OWNERS.md:
```diff
@@ -0,0 +1,8 @@
+
+See the [project Wiki TSC page](https://wiki.edgexfoundry.org/pages/viewpage.action?pageId=329436#TechnicalSteeringCommittee(TSC)-WorkingGroups) for information on the current EdgeX TSC and who occupies the role of Device Service Working Group chair.
+
+For a complete list of current committers see:  https://github.com/orgs/edgexfoundry/teams/device-bacnet-c-committers/members.
```
```suggestion
For a complete list of current committers see:  https://github.com/orgs/edgexfoundry/teams/device-can-committers/members.
```

vijay-hcl-epl replied:
Corrected

lenny-intel commented on README.md:
```diff
@@ -1,2 +1,109 @@
+## CAN Device Service Overview
+- Socketcan is used
+- The link - https://docs.kernel.org/networking/can.html#motivation-why-using-the-socket-api mentions why socketcan is preferred over character device-based driver
+- C – device-sdk-c is used. Version - 3.0.2
```
Remove version as it quickly becomes stall
```suggestion
- C – device-sdk-c is used. 
```

lenny-intel commented on README.md:
```diff
@@ -1,2 +1,109 @@
+Once the build is completed and interfaces are enabled, run the executable with the below commands
+
+export EDGEX_SECURITY_SECRET_STORE=false
+./build/release/device-can-c
```
```suggestion
./build/release/device-can
```

vijay-hcl-epl replied:
Corrected

lenny-intel commented on scripts/Dockerfile.alpine:
```diff
@@ -0,0 +1,53 @@
+RUN echo 'https://iotech.jfrog.io/artifactory/alpine-release/v3.16/main' >> /etc/apk/repositories
+RUN apk add --update --no-cache build-base git gcc cmake make linux-headers yaml-dev libmicrohttpd-dev curl-dev util-linux-dev ncurses-dev paho-mqtt-c-dev-1.3 hiredis-dev libcbor-dev iotech-iot-1.5-dev
+
+COPY scripts /device-can-c/scripts
```
Change all occurrences of  `device-can-c` to `device-can`

vijay-hcl-epl replied:
Corrected

lenny-intel commented on scripts/Dockerfile.alpine:
```diff
@@ -0,0 +1,53 @@
+COPY sample-profiles /sample-profiles
+
+LABEL license='SPDX-License-Identifier: Apache-2.0' \
+      copyright='Copyright (c) 2019-2022: IoTech Ltd'
```
Change copyright to HCL

vijay-hcl-epl replied:
Corrected

lenny-intel commented on scripts/Dockerfile.alpine-base:
```diff
@@ -0,0 +1,31 @@
+RUN echo 'https://iotech.jfrog.io/artifactory/alpine-release/v3.16/main' >> /etc/apk/repositories
+RUN apk add --update --no-cache build-base git gcc cmake make linux-headers yaml-dev libmicrohttpd-dev curl-dev util-linux-dev ncurses-dev paho-mqtt-c-dev-1.3 hiredis-dev libcbor-dev iotech-iot-1.5-dev
+
+COPY scripts /device-can-c/scripts
```
Change all occurrences of  `device-can-c` to `device-can`

vijay-hcl-epl replied:
Corrected

lenny-intel reviewed (changes requested):
I reviewed structure, build files, etc. Will leave C code to the C experts.

lindseysimple commented:
@vijay-hcl-epl Since this repo was not created from the template-repo, please  add a `.github` directory (see device-bacnet-c [repository](https://github.com/edgexfoundry/device-bacnet-c/tree/main/.github)) to align with EdgeX standards per @MightyNerdEric  's comment, thanks.

vijay-hcl-epl commented on Jenkinsfile:
```diff
@@ -0,0 +1,21 @@
+//
+// Copyright (c) 2019 Intel Corporation
```
Corrected

vijay-hcl-epl commented on Makefile:
```diff
@@ -0,0 +1,39 @@
+
+.PHONY: build test clean docker
+
+MICROSERVICES=build/release/device-can-c/device-can-c
```
Corrected

vijay-hcl-epl commented on Jenkinsfile:
```diff
@@ -0,0 +1,21 @@
+//
+
+edgeXBuildCApp (
+    project: 'device-can-c',
```
Corrected

vijay-hcl-epl commented on scripts/Dockerfile.alpine:
```diff
@@ -0,0 +1,53 @@
+#
+# Copyright (c) 2020 IOTech Ltd
```
Corrected

vijay-hcl-epl commented on scripts/Dockerfile.alpine-base:
```diff
@@ -0,0 +1,31 @@
+#
+# Copyright (c) 2020 IOTech Ltd
```
Corrected

vijay-hcl-epl commented:
@lenny-intel , @lindseysimple  Thank you for the review comments. I have corrected all the changes and added a new commit. 

Please let me know whether do I need to add the file Attribution.txt?

Thank You,
Vijay

lenny-intel reviewed (commented):
LGTM form structure POV will leave config and C code to the C device service experts.

lindseysimple commented:
> @lenny-intel , @lindseysimple Thank you for the review comments. I have corrected all the changes and added a new commit.
> 
> Please let me know whether do I need to add the file Attribution.txt?
> 
> Thank You, Vijay

Thanks @vijay-hcl-epl. Yes, we'll need to add Attribution.txt for the CAN device service as well. Please see [device-coap-c](https://github.com/edgexfoundry/device-coap-c/blob/main/Attribution.txt) for a reference.

lenny-intel commented:
> Please see [device-coap-c](https://github.com/edgexfoundry/device-coap-c/blob/main/Attribution.txt) for a reference.

@lindseysimple , the current C devices don't have proper Attribution.tx files. I have submitted issues to fix them.

@vijay-hcl-epl , please use https://github.com/edgexfoundry/device-mqtt-go/blob/main/Attribution.txt as a sample for the format. It should include all dependencies including the one from the C SDK.

lenny-intel commented:
@vijay-hcl-epl , I have reviewed the changes for all my comments and marked them as resolved. It is now up to the C experts to review the code.

FelixTing commented on src/c/device-can.c:
```diff
@@ -0,0 +1,375 @@
+
+	devsdk_service_free(service);
+	iot_data_free(driver_map);
+	free(impl);
```
add `free(canImpls;)` here

vijay-hcl-epl replied:
added free(canImpls)

FelixTing commented on src/c/device-can.h:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * @file
+ * @brief Defines common artifacts for the CoAP device service.
```
```suggestion
 * @brief Defines common artifacts for the CAN device service.
```

vijay-hcl-epl replied:
Corrected

FelixTing commented on scripts/Dockerfile.alpine:
```diff
@@ -0,0 +1,53 @@
+COPY LICENSE /.
+COPY Attribution.txt /.
+COPY res /res
+COPY sample-profiles /sample-profiles
```
The `sample-profiles` directory is not present.

FelixTing commented on scripts/Dockerfile.alpine:
```diff
@@ -0,0 +1,53 @@
+
+WORKDIR /device-can
+
+RUN /device-can/scripts/build_deps.sh
```
```suggestion
RUN /device-can/scripts/build_deps.sh 1
```

vijay-hcl-epl replied:
modified

vijay-hcl-epl commented on src/c/device-can.c:
```diff
@@ -0,0 +1,375 @@
+#define TIMEOUT "Timeout"
+
+can_driver *impl;
+iot_data_t *can_resp_data;
```
removed can_resp_data

vijay-hcl-epl commented on src/c/device-can.c:
```diff
@@ -0,0 +1,375 @@
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
```
removed unused headers

vijay-hcl-epl commented on src/c/device-can.c:
```diff
@@ -0,0 +1,375 @@
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/time.h>
+#include <sys/types.h>
```
removed unused headers

vijay-hcl-epl commented on src/c/device-can.c:
```diff
@@ -0,0 +1,375 @@
+#include <iot/data.h>
+#include "device-can.h"
+
+#include <stdarg.h>
+#include <unistd.h>
```
removed unused headers

vijay-hcl-epl commented:
> > Please see [device-coap-c](https://github.com/edgexfoundry/device-coap-c/blob/main/Attribution.txt) for a reference.
> 
> @lindseysimple , the current C devices don't have proper Attribution.txt files. I have submitted issues to fix them.
> 
> @vijay-hcl-epl , please use https://github.com/edgexfoundry/device-mqtt-go/blob/main/Attribution.txt as a sample for the format. It should include all dependencies including the one from the C SDK.

Thank you @lenny-intel & @lindseysimple. I have made the changes and submitted a commit

vijay-hcl-epl commented:
Thank you @FelixTing. I have made the changes and submitted a new commit.

lindseysimple reviewed (approved):
The project structure, `.github` directory content and `Addtribution.txt` all looks good to me, thanks Vijay.

FelixTing reviewed (approved):
LGTM. Will leave the device callback implementation, which is specific to the CANbus protocol, to CANbus specialists.

lenny-intel commented on Attribution.txt:
```diff
@@ -0,0 +1,29 @@
+edgexfoundry/device-sdk-c (Apache 2.0) https://github.com/edgexfoundry/device-sdk-c/v3.0.2
+https://github.com/edgexfoundry/device-sdk-c/blob/main/LICENSE
+
+edgexfoundry/go-mod-core-contracts (Apache 2.0) https://github.com/edgexfoundry/go-mod-core-contracts/v3
+https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/LICENSE
+
+edgexfoundry/go-mod-registry (Apache 2.0) https://github.com/edgexfoundry/go-mod-registry/v3
+https://github.com/edgexfoundry/go-mod-registry/blob/master/LICENSE
+
+edgexfoundry/device-sdk-go (Apache 2.0) https://github.com/edgexfoundry/device-sdk-go/v3
+https://github.com/edgexfoundry/device-sdk-go/blob/master/LICENSE
+
+edgexfoundry/go-mod-bootstrap (Apache 2.0) https://github.com/edgexfoundry/go-mod-bootstrap/v3
+https://github.com/edgexfoundry/go-mod-bootstrap/blob/master/LICENSE
+
+edgexfoundry/go-mod-secrets (Apache 2.0) https://github.com/edgexfoundry/go-mod-secrets/v3
+https://github.com/edgexfoundry/go-mod-secrets/blob/master/LICENSE
+
+edgexfoundry/go-mod-configuration (Apache 2.0) https://github.com/edgexfoundry/go-mod-configuration/v3
+https://github.com/edgexfoundry/go-mod-configuration/blob/master/LICENSE
+
+edgexfoundry/go-mod-messaging (Apache 2.0) https://github.com/edgexfoundry/go-mod-messaging/v3
+https://github.com/edgexfoundry/go-mod-messaging/blob/master/LICENSE
+
```
These are all got Go, not C. Please remove

vijay-hcl-epl replied:
Removed

lenny-intel commented on scripts/Dockerfile.alpine:
```diff
@@ -0,0 +1,52 @@
+RUN /device-can/scripts/build.sh 
+
+FROM alpine:3.16
+MAINTAINER iotech <support@iotechsys.com>
```
PLease update the Maintainer to be HCL

vijay-hcl-epl replied:
Updated

lenny-intel reviewed (approved):
LGTM

lenny-intel commented:
@lindseysimple , @FelixTing , I think we are ready for TSC vote to accept this new device service into EdgeX.

lindseysimple commented:
> @lindseysimple , @FelixTing , I think we are ready for TSC vote to accept this new device service into EdgeX.

Thanks @lenny-intel. I'll send a email for TCS vote.
